index.md

```diff
@@ -7,5 +7,44 @@ Best practices are difficult to practice without a good environment in which
+##<a name="environments"></a> Environments
+When developing and maintaining software several environments will be needed. Usually at least 4 environments are needed, each servering a different purpose:
+
+* Dev (development)
+    * Where developers have free rein(Admin) to do whatever they need to for development, such as install software for evaluation, test new architecures, or test production install instructions. 
+    * Code changes from Dev will most likely create new Int builds.
+    * each environment should be "sandboxed" from other environments to prevent any accidents to other resources.
+* Int (integration)
+    * Where developers combine code from everyone and test to make sure they merged correctly, and it works as expected. 
+    * Builds that pass in Int can be passed to Test for possible use in Prod.
+* Test
+    * Permissions and data are exact copies from Prod. 
+    * The data is updated on some predefined schedule determined by communication between developers and system admins. 
+    * End users and other testers verify bug fixes and new features. 
+    * Any build that passes in Test is eligible for promotion to Prod.
+* Prod (production)
+    * Where the built software is deployed for use by all users. 
+    * Actual data is entered, and software provides value. 
+    * Developers to have as little admin access as needed. 
+    * Prod also has a well defined process for deploying new test builds that have been promoted. See [Build Promotion] (#build-promotion) below.
+
+It should be noted that most (if not all) environments will need the full set of developer tools, this includes but is not limited to:
+
+* Compilier(s) / build tool(s)
+* Continuous integration server(s)
+* Debugger(s)
+
+## <a name="update-schedule"></a> Update Schedule
+Because updating software can affect multiple software and or systems, communication needs to agree upon for updates. It is also agree upon that critical patches to system can be done without notification before applying, and developers should design software to stop and restart at any time without reprocussions.
+
+## <a name="build-promotion"></a> Build promotion
+Every code check in creates a potentially deployable artifact due to Continuous integration server building after each code check in. Therefore a well defined and version controlled process for promoting a build from Dev(or continuous integration artifact)->Int->Test->Prod needs to exist.
+
+## <a name="self-provisioning"></a> Self Provisioning
+Developers can create and delete systems on-demand, with group alloction and quotas as needed to avoid over-commiting resources. These self-provisioned systems with documented exceptions for network configuration, autherntication, etc. All services should be developed using SSL. This requires a self-service system to issue certificates, so that applications can be developed and tested using the same configuration used in production.
```

Comment:
I think this should be broken into a list so some of the points can be expanded. “These self-provisioned systems with” should probably be something like “All self-provisioned systems should be identical to the production configuration except for”